### ISSUE TITLE
Reduce complexity

### ISSUE THREAD

**Issue by vilgotf** [PhasecoreX/docker-red-discordbot#24]
**Is your feature request related to a problem? Please describe.**
The image is large and over engineered

**Describe the solution you'd like**
Rewrite Dockerfile to be more minimal, preferably rebasing on alpine

**Describe alternatives you've considered**
I've begun using my [own image](https://github.com/vilgotf/Red-docker) which solves these issues. 

**Additional context**
opened in response to [Cog-Creators/Red-DiscordBot#4160](https://github.com/Cog-Creators/Red-DiscordBot/issues/4160#issuecomment-714375410)

**Comment by vilgotf** [comment]
| Name | Size
| --- | ---
| vilgotf/red-discordbot:audio | **307** MB
| vilgotf/red-discordbot:minimal (same concept as noaudio) | **139** MB
| phasecorex/red-discordbot:audio | **572** MB
| phasecorex/red-discordbot:noaudio | **365** MB

**Comment by PhasecoreX** [comment]
It's not possible to switch this image over to Alpine, as any redbot/cog dependencies compiled with Debian (what's currently being used, compiled with glibc) cannot run on Alpine (compiled with musl). Simply switching it over will break a large amount of bots currently using this image. Not worth a ~25MB size savings between the base images.

The main source of image size difference comes from the fact that my image keeps `build-essentials` installed, whereas your image removes `build-base`. Our images are solving the problem of installing redbot in different ways. Yours installs redbot as part of the build process, so no need for the `build-base` to stick around, whereas mine installs redbot during container startup, so it needs `build-essentials` present. There are benefits to my way, such as not waiting for the image owner to update their image when a new redbot version comes out. I think that's worth the size difference for being able to update redbot as soon as a new version is announced.

Also, having `build-essentials` stick around is great to support 3rd party cogs. I tried installing the notsobot cog with your image and it failed to install `numpy` and `pillow`[1]. Since your image is Alpine, no python wheels exist for it, and since you don't have `build-base`, those dependencies can't be compiled. Python wheels don't exist for other architectures (arm on the raspberry pi, for example), so I need `build-essentials` to be present to support those as well.

I ran the builds (which took like 4 minutes to complete, due to uvloop and friends needing to be compiled) without removing `build-base`. In case you wanted to support more 3rd party cogs, here are the resulting image sizes:

<table>
<tr>
	<td>
	<td>minimal/noaudio
	<td>audio
<tr>
	<td>phasecorex/red-discordbot
	<td>365 MB
	<td>572 MB
<tr>
	<td>vilgotf/red-discordbot (build-base)
	<td>338 MB (27 MB difference)
	<td>505 MB (67 MB difference)
<tr>
	<td>vilgotf/red-discordbot (no build-base)
	<td>139 MB
	<td>307 MB
</table>

As you can see, the image sizes are much closer. Granted, mine doesn't include redbot preinstalled, but again that's just the decision I made to allow for automatic redbot updates without relying on me to update the image.

TLDR, I can't switch to Alpine because it would break all the current users bots, and everything I have included in these images as far as I can tell is as minimal as possible for the amount of flexibility they support. If you think I can somehow reduce the size further with Debian, pull requests are appreciated!

[1]: The notsobot cog will install on mine just fine, but it will not load since ImageMagick/Wand is missing. I only use notsobot as an example because it had two common packages that wouldn't install that are used in some other 3rd party cogs.

**Comment by vilgotf** [comment]
Installing new versions of redbot inside of the container breaks the immutability consensus of how Containers should behave.
New versions should be built and tagged by CI so I disagree with the premiss of atomic images leading to update delays.
One can moreover just create more tags whenever needed that include the dependencies of popular cogs or simply extend the base image by hand (dockerhub allows automatic image rebuilding if baseimage is updated).

**Comment by vilgotf** [comment]
Your repo could for instance support Alpine images in addition to the ones based on Debian (and we would both be happy?)

**Comment by PhasecoreX** [comment]
This repo isn't tied to the Red-DiscordBot repo in any way. If there is an emergency update to redbot at 3am, I'm not able to update my images until probably 5pm since I work. Or, if I'm on vacation for a week, those images aren't getting updated anytime soon. You can look at the container as an environment that has everything in it that you need to run redbot. You would just supply the redbot for it to run (or, it downloads it for you, in a venv in your data volume).

If compiled cog dependencies weren't a thing, offering an Alpine version would be easy. Alpine and non Alpine tags being offered on an image to me means you can freely switch between them, because everything compiled is in the image itself. Redbot is different, where it can install whatever python packages it wants, and thus can compile anything it wants. Those compiled cogs go into the users data directory (specifically the Downloader cogs `lib` folder). So you can't just freely switch to Alpine and back, as you'll have a bunch of dependencies compiled the wrong way.

You could of course preinstall redbot and a bunch of popular dependencies, but then your image gets big and there's always going to be that one other dependency you missed and need to add. Or you can just let the container do all the downloading and compiling when and if it needs to, at the expense of a larger image size. I chose the second option.

In terms of DockerHub, it only does amd64 builds, so it's a no-go for multi-arch (at the moment). Auto builds triggered by a base image updating only works if the base image isn't a core library image (like `python:3.8-alpine` or `python:3.8-slim`).

**Comment by vilgotf** [comment]
Sure but there's other CI that does support multiarch and official aswell as nonofficial baseimages (like Github actions)

**Comment by Drapersniper** [comment]
> opened in response to Cog-Creators/Red-DiscordBot#4160

I would like to point out that this is Phasecore personal Repository. This image is not officially supported, and No user should demand feature from Phase due to them seeing that issue.
